### PR TITLE
Training to predict x0 in training example

### PR DESCRIPTION
**Pull request by lukovnikov**
changed training example to add option to train model that predicts x0 (instead of eps), changed DDPM pipeline accordingly

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
@patil-suraj @anton-l could you take a look here? :-)

**Review by patil-suraj — APPROVED** [review]
Thanks a lot @lukovnikov ! The PR looks good.  Out of curiosity, what does SNR weighting do ? And is it only used when predicting x0 ?

**Comment by lukovnikov** [comment]
@patil-suraj I'm not sure it's the best name for it, but it's the weighting that is equivalent to having equal weights for all time steps in eps-prediction mode, after you transform the equations to x0-prediction mode. I followed https://arxiv.org/pdf/2202.00512v2.pdf but maybe it's older.

**Comment by lukovnikov** [comment]
Btw, do you have any FID numbers to check how good the model gets? I ran a quick training in x0-prediction mode and drew 8k samples, and against the 8k original samples (after cropping), it gives me a FID of around 50. The images look ok but could be much better. I haven't had time to generate for the eps-prediction model yet so wanted to ask.

**Review by anton-l — APPROVED** [review]
Nice addition @lukovnikov, thank you! 
Before merging, the PR needs to pass the codestyle checks: just run `pip install diffusers[quality]` and `make style` from the repository root to reformat everything :) 

Regarding FID: we don't have the numbers yet, and the metric PR in our `evaluate` library is still pending: https://github.com/huggingface/evaluate/issues/276 If you want to implement it, that would be awesome!

**Comment by lukovnikov** [comment]
@anton-l 
Great, I'll try fix code style this week.
Regarding evaluation, can't we just write the images to folders (both from the dataset and from the model) and run one of the existing tools from command-line (torch-fidelity or pytorch-fid), at least for now? I'm always skeptical and paranoid of reimplementations where I can't see both image sets.

**Comment by anton-l** [comment]
@lukovnikov yes, saving the eval predictions at the end of training a good option too! Let's do it in a follow-up PR :rocket: 

The style checks pass now, merging :hugs:

**Comment by lukovnikov** [comment]
Huh, i was looking but some pytorch checks were failing although it looked like it was something about stable diffusion.

**Comment by anton-l** [comment]
Yeah, those failing tests were the Apple M1 issues in our testing setup, all good now :)
